### DISCUSSION TITLE
GCP Default Application Credentials (authenticate without JSON key)

### DISCUSSION THREAD

**Discussion by hermanbanken** [fastlane/fastlane#20022] [General]
https://github.com/fastlane/fastlane/blob/master/supply/lib/supply/client.rb expects hardcoded JSON key credentials for GCP. However, Google advises against this, and there is now a much better solution: using the default Google Application Credentials, and using Workload Identity Federation (OIDC) on GitHub/ci-cd-platforms.

https://cloud.google.com/blog/products/identity-security/enabling-keyless-authentication-from-github-actions

![](https://storage.googleapis.com/gweb-cloudblog-publish/images/2_GitHub_Actions.max-1100x1100.jpg)

I want to set this up, but due to how fundamentally hardcoded the assumption of having a JSON Key is - and my absent Ruby skills - I'm a bit lost.

The crux is allowing to setup a Supply client without explicit credentials, defaulting to any configured `gcloud` credentials:

```diff
-Google::Auth::ServiceAccountCredentials.make_creds(json_key_io: service_account_json, scope: self.class::SCOPE)
+Google::Auth::ServiceAccountCredentials.make_creds(scope: self.class::SCOPE)
```

Note that if you do this, you can rely on [things like](https://cloud.google.com/docs/authentication/getting-started):
- GCP Metadata API (for CloudBuild and GCP VM)
- `GOOGLE_APPLICATION_CREDENTIALS` like 
    ```bash
    export GOOGLE_APPLICATION_CREDENTIALS="/home/user/Downloads/service-account-file.json"
    ```
- Workload Identity Federation from GitHub, see https://github.com/google-github-actions/auth

--------

Would require some change like this:
https://github.com/fastlane/fastlane/compare/master...hermanbanken:patch-2

I have no idea how to test this however.

**Comment by hermanbanken** [comment]
Perhaps the setting `create_credentials_file` of https://github.com/google-github-actions/auth can be used instead. This generates a JSON file with is compatible with SUPPLY_JSON_KEY_DATA and `Google::Auth::ServiceAccountCredentials`, but does not contain any private information and will be generated automatically.

I'll try this. If it works it makes sense to update the documentation to reflect this.

Update: does not work due to error `read_json_key': [!] missing client_email (RuntimeError)` by the Google Auth library.

**Comment by hermanbanken** [comment]
Maybe we can use [`Signet::OAuth2::Client.new(access_token: token)`](https://github.com/googleapis/google-auth-library-ruby/commit/9d7051cff4d5e191a5d6756a068e8be539934f0d#diff-1f34c7b8878c907d82ca7a929b0bc17c5e5e2305f7b046b2e594d10e0477e6aaR100-R102):

```ruby
    signet = Signet::OAuth2::Client.new access_token: token # Client#needs_access_token? will return false
    creds = Google::Auth::Credentials.new signet
```

**Reply by hermanbanken** [reply]
https://github.com/fastlane/fastlane/pull/20034

**Comment by DevNico** [comment]
Is this actively being looked into? We've migrated all our authentication to using WIF but fastlane is our last blocker.

**Reply by DevNico** [reply]
We'd like to use it for match certs stored in Cloud Storage

**Reply by hermanbanken** [reply]
I don't think so. If you have some Ruby skills it shouldn't be too hard to build the support, but I unfortunately don't have that experience.

**Reply by hermanbanken** [reply]
Maybe PR https://github.com/fastlane/fastlane/pull/20319 works for you?

**Reply by DevNico** [reply]
Probably not since that seems to only work on the supply action. I'd need support for this in the match/sync_code_signing action. I tried looking into it myself and I think support has to land in https://github.com/googleapis/google-auth-library-ruby first but my Ruby knowledge is limited to writing the lanes in our Fastfiles.

**Reply by hermanbanken** [reply]
There is a PR in that one for this: https://github.com/googleapis/google-auth-library-ruby/pull/386 but it is pending review.

**Reply by trenslow** [reply]
did anyone ever figure out how to do this? it seems like the ruby google auth library supports WIF now: https://github.com/googleapis/google-auth-library-ruby/pull/418., but its not clear to me if fastlane will support the credentials obtained via WIF in a previous action step

**Reply by amadeu01** [reply]
did anyone solve this for fastlane ?

**Reply by trenslow** [reply]
looks like someone is trying to solve it https://github.com/fastlane/fastlane/pull/27949

**Reply by trenslow** [reply]
coming back to say that this works since fastlane 2.232.0 was released, heres [the commit](https://github.com/fastlane/fastlane/commit/5c8853f7c6d5f8b1ed946201cbbdf78911f03064) 

heres an example thats tried and tested on my side:

```yaml
      - name: Authenticate with Google Cloud
        id: auth-gcp
        uses: google-github-actions/auth@7c6bc770dae815cd3e89ee6cdf493a5fab2cc093 # v3.0.0
        with:
          workload_identity_provider: ${{ vars.GCP_WI_PROVIDER }}
          service_account: ${{ vars.GCP_WI_SVC_ACCT }}
          create_credentials_file: true
      - name: Upload to Google Play
        run: |
          fastlane supply \
              < extra config here >
              --json_key '${{ steps.auth-gcp.outputs.credentials_file_path }}'
```

so this discussion may be able to be closed

it unfortunately doesnt use local developer machine ADC, since the local credentials are of type `authorized_user`, and [the code](https://github.com/fastlane/fastlane/blob/09e0f765567579826917d6cf22e2350c159a7a2e/supply/lib/supply/client.rb#L59) doesnt allow for this type. but its a major step in the right direction
